### ISSUE TITLE
psycopg2 error deploying to staging environment via gitlab pipeline

### ISSUE THREAD

**Issue by zkwurst** [tiangolo/full-stack-fastapi-postgresql#136]
I've really enjoyed the time I've spent working with this repo and want to ask for some help deploying to my remote server.

**Background:**
The default setup for this repo has everything running in a single postgres service called "postgres" with a db called "app". My setup has everything running in a postgres service called "srv_app" with the db called "db_app". I also want to connect additional postgres servers running in the same environment that store streaming data.  I've been able to get everything running on my local machine at localhost including the additional streaming servers.

**Problem:**
Everything breaks when I deploy to my staging env. I can't even get the single "srv_app" container to communicate with the "db_app" container. I'm deploying to a DigitalOcean server running Docker Swarm via GitLab CI/CD (followed the Docker Swarm Rocks tiangolo docs) with a container registry that stores each of the containers not built from images. Everything in the GitLab pipeline tests and builds fine and the staging url opens the default login page like it's supposed to, but returns an incorrect password or username error when I try to login.

When I ssh into the server and inspect the container logs, the celeryworker and the db_app containers say that psycopg2 can't connect to the database and keeps returning this error until it times out:

```bash
ERROR:__main__:(psycopg2.OperationalError) could not translate host name "srv_app" to address: Name or service not known
```

The initdb scripts on container startup access postgres just fine, but when psycopg2 tries to connect from the db_app container, it fails trying to use whatever is in the `POSTGRES_SERVER` variable. I've tried everything I can find online to connect the containers with docker compose networks, declaring ports, changing postgres credentials for different environments, but I can't figure out how to get 'db_app' to communicate with the 'srv_app' postgres service.

To test the connections (how to do this securely is another issue), I'm just passing .env files along with the containers through the ci/cd pipeline.  ie. `env_srv_app.env`:
```env
POSTGRES_SERVER=srv_app
POSTGRES_USER=postgres
POSTGRES_PASSWORD=postgres
POSTGRES_DB=postgres
```

The sqla uri and alembic function both work locally:
```python
# core config.py
SQLALCHEMY_DATABASE_URI = (
    f"postgresql://{POSTGRES_USER}:{POSTGRES_PASSWORD}@{POSTGRES_SERVER}/{POSTGRES_DB}"
)

# alembic .env file
def get_url():
    user = os.getenv("POSTGRES_USER", "")
    password = os.getenv("POSTGRES_PASSWORD", "")
    server = os.getenv("POSTGRES_SERVER", "")
    db = os.getenv("POSTGRES_DB", "")
    return f"postgresql://{user}:{password}@{server}/{db}"
```

Has anyone else run into something like this? Any ideas on how to get this pipeline working? I'd be thrilled to get the whole stack running through the ci/cd pipeline. Happy to provide any other information about my setup and any help would be most appreciated. Thank you!

**Comment by zkwurst** [comment]
#116 #105 #78 seem related to my issue. Haven't found any fixes in those issues yet.

**Comment by zkwurst** [comment]
I finally got everything working. Wanted to close this out and leave a few notes:

**TL;DR** the `db` or `srv_app` image wasn't properly replicating on my remote server as a service because the postgres image was failing to access a volume on another container.

During local builds, I was passing a directory of .sh scripts into the postgres image `/docker-entrypoint-initdb.d` which would execute during `docker-compose up`. For the staging pipeline, I copied that directory into the Dockerfile for the `backend` (or `db_app` container in my case) and tried to give postgres access to it by declaring that volume in the postgres image build. When I removed the line declaring that volume, the postgres image would build properly on the staging server. After many attempts at shared networks, volumes, containers, etc, I stopped trying to figure it out and am now running .py scripts from the `db_app` container to set everything up after the postgres server builds.

The error output was pretty limited for this. Besides the connectivity errors, the only indication about the cause was the `0/1` under the replicated column for postgres when running `docker service ps` on the remote staging server. I remember seeing a FR somewhere in a gitlab or postgres forum asking to improve the error or logging output of the gitlab runner or docker service when something failed to replicate like this, but I can't seem to find it again. Will update if I come across it.

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue! :+1: 

I'm glad you solved it.

In any case, if you need to add custom scripts to your Postgres container you would probably have to create a custom image with it's own `Dockerfile` adding those files.

**Comment by zkwurst** [comment]
My pleasure, @tiangolo. Thanks for all you do. Glad to support this work and community however I can.
